### ISSUE TITLE
PDF Viewer seems to cache documents based on fingerprint hash

### ISSUE THREAD

**Issue by halufa** [stephanrauh/ngx-extended-pdf-viewer#814]
I've a system that fills PDF form files and store them on separate PDF files.
Then I have a another system that displays those PDF files. Everything works correctly on the first opened file. The following files display as the first one. If I click the print button the print preview displays all the files correctly.
If I try to view another file with diferent fingerprint hash it also works.

Is this a bug? Is there a way to disable this caching behaviour?

By fingerprint I mean the hash displayed at the console: PDF 41df0efab365433c88a12ff042b084f3 [1.7 pdf-lib (https://github.com/Hopding/pdf-lib) / Adobe Illustrator CC 23.0 (Macintosh)] (PDF.js: 2.9.149) modified by ngx-extended-pdf-viewer)

This happens in all the browsers I've tested and version 8.2, 8.3 and 9.0.2 of gx-extended-pdf-viewer.

Thanks

**Comment by stephanrauh** [comment]
Oh, that's interesting. Your PDF files are identical, so the fingerprints are identical, too. But maybe the fingerprints should be different because there is a difference: the form data saved along with the PDF file is different.

@timvandermeij Is this something you'd like to take care of? I know pdf.js is "only" supposed to be the PDF viewer of Firefox, but even so, I imagine Firefox refuses to open a different PDF file with the same fingerprint but different form data. So is it my job or your job to help Paulo?

Just for the sake of completeness: the fingerprint stems from this line (slightly modified from pdf.js, but you'll recognize it):
```typescript
console.log(
        `PDF ${pdfDocument.fingerprint} [${info.PDFFormatVersion} ` +
          `${(info.Producer || "-").trim()} / ${(info.Creator || "-").trim()}] ` +
          `(PDF.js: ${version || "-"}` +
          `${
            this.pdfViewer.enableWebGL ? " [WebGL]" : ""
          }) modified by ngx-extended-pdf-viewer)`
      );
```

**Comment by timvandermeij** [comment]
I'm not sure I fully understand the question here. I don't see how the PDF files can be equal and have different form data since the form data is in the file, which would make the files themselves different and they should therefore get a different fingerprint/hash. It's also not possible to verify this without the actual files, but the fingerprint logic lives here:

https://github.com/mozilla/pdf.js/blob/9de0916fd473244cd23133c79d5aaa52ee6ce707/src/core/document.js#L1098-L1122

If the ID field in the trailer is equal while the form data is different, that might actually be a PDF generator bug. If there is no ID we fingerprint the first 1024 bytes of the file, which should really be unique enough.

In short, if there is a bug in this logic we're happy to fix it in PDF.js, but we can only really tell if the PDF files are provided.

**Comment by stephanrauh** [comment]
Thanks, Tim! @halufa I'm afraid now it's your turn. Would you mind to send me two PDF files with different form data but identical fingerprints?

**Comment by halufa** [comment]
Yes @stephanrauh. Is there a way to send the files to you directly? I don't want them shared publicly.

**Comment by stephanrauh** [comment]
No problem. Send the files to webmaster1 at beyondjava.de.

**Comment by halufa** [comment]
Email sent. Thank you.

**Comment by stephanrauh** [comment]
I can reproduce the bug. Obrigado!

**Comment by stephanrauh** [comment]
Your base PDF file has an ID in the trailer directory. I assume that's a data structure created by the tool that generated your PDFs. In general, that's a good choice to create a fingerprint. Unfortunately, the trailer ID doesn't change if you fill in form data. That's why the fingerprints are identical.

I'm suppose you, Paulo, can't do anything about the trailer ID. So let's start to find "plan B".

**Comment by stephanrauh** [comment]
It doesn't seem to be easy. I'm sure there's a way, but currently I don't know where and how the document is cached.

If you're ready to implement a brute force workaround: a full-page reload does the trick. But I'm sure there are more elegant options!

**Comment by stephanrauh** [comment]
As it turns out, the bug was something completely different. The form data is rendered as a independent layer, so it's OK to cache based on a fingerprint that doesn't change when form data changes. The two-way data binding I've implemented for Angular caused the bug. After removing my copy of the form data before opening a new document everything works fine again.

**Comment by stephanrauh** [comment]
The bugfix has landed (version 9.0.3).

Enjoy!
Stephan
